pkg/auth/oauth/handlers/grant.go

```diff
@@ -140,32 +141,41 @@ func (g *autoGrant) GrantNeeded(user user.Info, grant *api.Grant, w http.Respons
 }
 
 type redirectGrant struct {
-	url string
+	subpath string
 }
 
-// NewRedirectGrant returns a grant handler that redirects to the given URL when a grant is needed.
+// NewRedirectGrant returns a grant handler that redirects to the given subpath when a grant is needed.
 // The following query parameters are added to the URL:
 //   then - original request URL
 //   client_id - requesting client's ID
 //   scopes - grant scope requested
 //   redirect_uri - original authorize request redirect_uri
-func NewRedirectGrant(url string) GrantHandler {
-	return &redirectGrant{url}
+func NewRedirectGrant(subpath string) GrantHandler {
+	return &redirectGrant{subpath}
 }
 
 // GrantNeeded implements the GrantHandler interface
 func (g *redirectGrant) GrantNeeded(user user.Info, grant *api.Grant, w http.ResponseWriter, req *http.Request) (bool, bool, error) {
-	redirectURL, err := url.Parse(g.url)
-	if err != nil {
-		return false, false, err
-	}
-	redirectURL.RawQuery = url.Values{
-		"then":         {req.URL.String()},
-		"client_id":    {grant.Client.GetId()},
-		"scope":        {grant.Scope},
-		"redirect_uri": {grant.RedirectURI},
-	}.Encode()
-	http.Redirect(w, req, redirectURL.String(), http.StatusFound)
+	_, lastSegment := path.Split(req.URL.Path)
+
+	// Make our "then" URL a relative backstep
+	reqURL := &(*req.URL)
+	reqURL.Host = ""
+	reqURL.Scheme = ""
+	reqURL.Path = path.Join("..", lastSegment)
```

Comment:
you misunderstand, it gets rid of the trailing slash in the browser... if you just redirect to `.`, you'd get `/oauth/authorize/`